src/Microsoft.VisualStudio.Threading/ReentrantSemaphore.cs

```diff
@@ -0,0 +1,461 @@
+            protected override void ThrowIfFaulted()
+            {
+                Verify.Operation(!this.faulted, Strings.SemaphoreMisused);
+            }
+        }
+
+        /// <summary>
+        /// An implementation of <see cref="ReentrantSemaphore"/> supporting the <see cref="ReentrancyMode.Freeform"/> mode.
+        /// </summary>
+        private class FreeformSemaphore : ReentrantSemaphore
+        {
+            /// <summary>
+            /// The means to recognize that a caller has already entered the semaphore.
+            /// </summary>
+            private readonly AsyncLocal<Stack<AsyncSemaphore.Releaser>> reentrantCount = new AsyncLocal<Stack<AsyncSemaphore.Releaser>>();
+
+            /// <summary>
+            /// Initializes a new instance of the <see cref="FreeformSemaphore"/> class.
+            /// </summary>
+            /// <param name="initialCount">The initial number of concurrent operations to allow.</param>
+            /// <param name="joinableTaskContext">The <see cref="JoinableTaskContext"/> to use to mitigate deadlocks.</param>
+            internal FreeformSemaphore(int initialCount, JoinableTaskContext joinableTaskContext)
+                : base(initialCount, joinableTaskContext)
+            {
+            }
+
+            /// <inheritdoc />
+            public override async Task ExecuteAsync(Func<Task> operation, CancellationToken cancellationToken = default)
+            {
+                Requires.NotNull(operation, nameof(operation));
+                this.ThrowIfFaulted();
+
+                Stack<AsyncSemaphore.Releaser> reentrantStack = this.reentrantCount.Value;
+                if (reentrantStack == null || reentrantStack.Count == 0)
+                {
+                    this.reentrantCount.Value = reentrantStack = new Stack<AsyncSemaphore.Releaser>(capacity: 2);
+                }
+
+                Func<Task> semaphoreUser = async delegate
+                {
+                    using (this.joinableTaskCollection?.Join())
+                    {
+                        AsyncSemaphore.Releaser releaser = reentrantStack.Count == 0 ? await this.semaphore.EnterAsync(cancellationToken).ConfigureAwait(true) : default;
+                        lock (reentrantStack)
+                        {
+                            reentrantStack.Push(releaser);
+                            releaser = default; // we should release whatever we pop off the stack (which ensures the last surviving nested holder actually releases).
+                        }
+
+                        try
+                        {
+                            await operation().ConfigureAwaitRunInline();
+                        }
+                        finally
+                        {
+                            lock (reentrantStack)
+                            {
+                                releaser = reentrantStack.Pop();
+                            }
+
+                            DisposeReleaserNoThrow(releaser);
```

Comment:
>DisposeReleaserNoThrow(releaser); [](start = 28, length = 33)

You might have pushed null (default) above.  Need a null check here or defend against null in ```DisposeReleaserNoThrow```

Reply:
Does this imply a test hole?

---
In reply to: [194844826](https://github.com/Microsoft/vs-threading/pull/296#discussion_r194844826) [](ancestors = 194844826)

Reply:
The `Releaser` type is a struct, so it can't be null, and in fact is safe to Dispose on its default instance.
That in fact is how this stack works: only the original Releaser is meaningful -- the rest or `default` instances that no-op when disposed.

Reply:
Regarding the test hole, no. We have tests that include default instance releasers being pushed and popped. That's how reentrancy works and we have several tests for that.